Blog/ambient mc (#16750)
istio/istio.io: content/en/blog/2025/ambient-multicluster/index.md

Reviewer: I haven't done a full review yet but I'm happy with the level the latest draft is at!

---
title: Introducing multicluster support for ambient mode (alpha)

> description: Introducing multicluster support for ambient mode (alpha).

Reviewer: These should be different from each other; description is probably optional

Reviewer: I recall a dumb lint rule that will complain if one of these doesn't end with a period

date: 2025-08-04
attribution: Jackie Maertens (Microsoft), Keith Mattix (Microsoft), Mikhail Krinkin (Microsoft), Steven Jin (Microsoft)
keywords: [ambient,multicluster]
---

Multicluster has been one of the most requested features of ambient -— and as of Istio 1.27, it is available in alpha status!

> We sought to capture the benefits and avoid the complications of multicluster architectures using the same modular design that ambient users love.

Reviewer 2: Not sure if this is better?
```suggestion
We sought to capture the benefits and avoid the complications of multicluster architectures while using the same modular design that ambient users love.
```

## Key Components

Ambient multicluster extends ambient with new components and minimal APIs to
securely connect clusters using ambient's lightweight, modular architecture.
It builds on the namespace sameness model -- a service in namespace `foo` in one cluster is treated as the same logical service as `foo` in another --
so services keep their existing DNS names across clusters, allowing you to control cross-cluster communication without changing application code.

### East-West Gateways

Each cluster has an east-west gateway with a globally routable IP acting as an entry point for cross-cluster communication.
A ztunnel connects to the remote cluster's east-west gateway, identifying the destination service by its namespaced name.
The gateway then load balances the connection to a local pod.
Using the gateway’s routable IP removes the need for inter-cluster routing configuration,
and addressing pods by namespaced name rather than IP eliminates issues with overlapping IP spaces.
Together, these design choices enable cross-cluster connectivity without changing cluster networking or restarting workloads,
even as clusters are added or removed.

> ### Double HBONE

Reviewer: That's the section I was missing from the documentation :)

Although the default value is straightforward, the API is flexible and can express complex conditions using a mix of ANDs and ORs.

By default, ztunnel load balances traffic uniformly across all endpoints --even remote ones--,
but is configurable through the service's `trafficDistribution` field to only cross cluster boundaries when there are no local endpoints.
Thus, users have control over whether and when traffic crosses cluster boundaries with no changes to application code.

## Limitations and Roadmap

Although the current implementation of ambient multicluster has the foundational features for a multicluster solution,
there is still a lot of work to be done.
We are looking to improve the following areas

* Service and waypoint configuration must be uniform across all clusters.
* No cross-cluster L7 failover (L7 policy is applied at the destination cluster).
* No support for direct pod addressing or headless services.
* Support only for multi-primary deployment model.
* Support only for one network per cluster deployment model.

We are also looking to improve our reference documentation, guides, testing, and performance.

> If you would like to try out ambient multicluster, please follow [this guide](TODO).

Reviewer 3: Just making a note that this will need to b e changed before merging. I think it's waiting on #16709

keywords: [ambient,multicluster]
---

Multicluster has been one of the most requested ambient features — and as of Istio 1.27, it's now available.
We sought to capture the benefits and avoid the complications of multicluster architectures using the same modular design that ambient users love.
While still in alpha, this release delivers the core functionality of a multicluster mesh and lays the groundwork for a full feature set in upcoming releases.

## Multicluster's Many Benefits (and Challenges)

Multicluster architectures increase outage resilience, shrink the blast radii,
ease adoption of data residence policies, and simplify cost tracking.
That said, integrating multiple clusters poses connectivity, security, and operation hurdles.

In a single Kubernetes cluster, every pod can directly connect to another pod via a unique pod IP or service VIP.
We lose these guarantees when we start thinking of multicluster architectures.
IP address spaces of different clusters might overlap.
Even if they didn't, nodes in one cluster would not know how to route traffic from one cluster to another.

Establishing cross-cluster connectivity also presents security challenges.

> Cross-cluster connectivity means that pod-to-pod traffic can leave cluster boundaries -- and that pods may accept connections from outside the cluster.

Reviewer 4: We don't allow users to write policies based on the cluster the traffic resides in so I'd rephrase to talk about the complexity of deriving an identity model across clusters

Author: I'm interpreting the sentence as "Cross-cluster connectivity means that traffic is (often crossing the cluster) without going through an ingress gateway."

I don't really want to highlight differentiate the difference between ingress and e/w gateways. They are both proxies that can apply some sort identity verification and route traffic to their destinations within the cluster. I don't think that, from an identity point of view, they are inherently different (though in practice they are used for different purposes and configured with very different policies). But I do like more explicitly mentioning identity enforcement at the edge of the cluster. I would just frame it in terms of the attacker

Without care, an attacker could connect to a vulnerable pod, or sniff unencrypted traffic.

For a multicluster solution to be viable, it must at least securely connect clusters, and do so
through APIs that are simple enough to keep pace with ever-changing environments.

## Key Components.

> Ambient multicluster extends ambient with new components and minimal APIs to

Reviewer 4: Maybe add a phrase about namespace sameness here

securely connect clusters using the same lightweight, modular architecture.

### East-West Gateways

Each cluster deploys an east-west gateway with a globally routable IP that acts as an entrypoint for cross-cluster communication.

> The east-west gateways are configured using GatewayAPI and controlled by istiod.

Reviewer 4: ```suggestion
The east-west gateways are configured using Gateway API and controlled by istiod.
```

Reviewer 4: I also don't know how important it is to include this sentence. What is the user gaining from it?

> A ztunnel communicates across clusters by connecting to the remote cluster's east-west gateway and sending the destination service FQDN.

Reviewer 4: The FQDN format is a bit too in the weeds IMO

Author: agreed. Using service name instead. Its more than anything that we don't send with service vip.

The east-west gateway will then forward the connection to a cluster-local pod of its choosing.
As such, overlapping IP spaces are of no concern because we never directly address a pod in a remote cluster.
Ambient multicluster achieves cross-cluster connectivity without changes to cluster networking configuration.

> We can achieve this connectivity using only ambient and declarative APIs.

Reviewer 4: Again, don't feel like this adds a ton

Author: this line or paragraph?

Author: probably line.

Reviewer 4: Yep line

There is no need to restart workloads, manage IP address spaces, or configure routing tables.

> 

Reviewer 4: Is this a good place to note that headless services or direct pod addressing aren't currently supported?

Author: I'll put it in limitations

### Double HBONE

Ambient Multicluster uses nested [HBONE](https://istio.io/latest/docs/ambient/architecture/hbone/) connections to secure traffic traversing cluster boundaries while preserving ambient's strong security.

> An outer HBONE connects the source ztunnel to its east-west gateway while an inner HBONE tunnel extends the connection to the destination.

Reviewer 4: Is this sentence necessary when the next two break down the inner and outer tunnels again?

The outer HBONE connection encrypts cross-cluster traffic, encrypts the destination service FQDN, and allows the source ztunnel and east-west gateway to verify each other's identity.
The inner HBONE connection encrypts traffic end-to-end, which allows the source ztunnel and destination ztunnel to verify each other's identity.
Put together, the two HBONE layers stop unauthenticated access, protect against data sniffing, and allow identity verification at every step.
At the same time, the HBONE layers allow ztunnel to effectively reuse cross-cluster connections, minimizing TLS handshakes.

> One drawback is that we encrypt application data twice (once for the outer HBONE and once for the inner HBONE).

Reviewer 4: These two sentences are two deep in the weeds, remove

We found this to be an acceptable drawback because it allows us to stick with open standards, and we expect the extra encryption to be negligible compared to the cost of sending data across clusters.

{{< image link="./mc-ambient-traffic-flow.png" caption="Istio ambient multicluster traffic flow" >}}

### Service discovery and scope

Once we have securely connected our clusters, we enable cross-cluster communication for a service by marking it global.
When a service becomes global, istiod will configure east-west gateways to accept and route traffic destined to the global service.
Istiod will also read remote apiservers and configure ztunnel with the number of pods for the global service per remote cluster.
Ztunnel can then proxy traffic to the global service across clusters.

Mesh administrators define the label-based criteria for global services via the `ServiceScope` API,
and app developers opt into global behavior by labeling their services accordingly.
The default `ServiceScope` is

{{< text yaml >}}

> =======

Reviewer 4: This gets rendered strangely; maybe remove

serviceScopeConfigs:
  - servicesSelector:
      matchExpressions:
        - key: istio.io/global
          operator: In
          values: ["true"]
    scope: GLOBAL
{{< /text >}}

meaning that any service with the `istio.io/global=true` label is global.
Although the default value is straightforward, the API is flexible and can express complex conditions using a mix of ANDs and ORs.

By default, ztunnel will load balance traffic uniformly across all endpoints --even remote ones--, but this can be configured using the service's `trafficDistribution` field to only cross cluster boundaries when there are no local endpoints.
Thus, users have control over whether and when traffic crosses cluster boundaries.

> ## Limitations and Roadmap

Reviewer 4: General feedback on this section: limitations should be short and digestible. Readers should be able to know if this is something they can use within 5 seconds of getting here. Less words, more bullet points and links

Although the current implementation of ambient multicluster has the foundational features for a multicluster implementation,
there is still a lot of work to be done.

For example, currently, we require that global services, attached waypoints, and `ServiceScope` configuration have uniform configuration across all clusters.

> This greatly simplified our alpha implementation, but we are looking to allow for more configuration skew.

Reviewer 4: Remember: we're not giving a report or presenting an academic paper. The readers don't really care about what our experience was building this; they want to know how we're adding value to them.

Similarly, waypoints and L7 policy enforcement have proven difficult since different clusters might have different policies.
In our alpha implementation, if a service has a waypoint, it will go through said waypoint in the destination cluster.
This reduces unexpected surprises by enforcing the destination cluster's L7 authorization policy, but remove the ability to perform L7 cross-cluster failover.
Eventually, we would like to apply L7 policy in both the source and destination cluster.
We are also looking to improve our reference documentation, guides, testing, and performance.
Currently, we only support a multi-primary deployment model with a single network per cluster, but would eventually want to support other cluster and network models.

If you would like to try out ambient multicluster, please follow [this guide](TODO).
Since many details are in discussion, we would love to hear any of your thoughts, comments, and use cases.

> You can find ways to reach us on the [Istio community page](https://istio.io/latest/about/community/).

Reviewer 4: Maybe reduce their clicks by just linking to slack/github? This is also a good spot to stress that this is alpha and NOT ready for production. "If you find a bug, please file it on GitHub" is a good way to end that

---

> title: Introducint Ambient Multicluster

Reviewer: Yep - ambient mode isn't a Brand Name. I'll do a pass for such things.

Put together, the two HBONE layers stop unauthenticated access, protect against data sniffing, and still allow ztunnel to verify the destination’s identity.
At the same time, it allows ztunnel to effectively reuse cross cluster connections, minimizing TLS handshakes.

The one drawback is that we encrypt application data twice (once for the outer HBONE and once for the inner HBONE).
We found this to be an acceptable drawback because it allows us to stick with open standards, and we expect the extra encryption to be negligible compared to the cost of sending data across clusters.

{{< image link="./mc-ambient-traffic-flow.png" caption="Istio Ambient Multicluster traffic Flow" >}}

### ServiceScope API

Once clusters are securely connected, marking services as global to allow cross cluster communication,
the `ServiceScope` API allows mesh administrators to mark which combinations of labels make a service global,
and app developers can label their services accordingly.
A global service is one has endpoints in all clusters and can be accessed from any cluster.
The default `ServiceScope` is

> {{< text yaml >}}
>   serviceScopeConfigs:
>     - servicesSelector:
>         matchExpressions:
>           - key: istio.io/global
>             operator: In
>             values: ["true"]
>       scope: GLOBAL
> {{< /text >}}

Reviewer: ```suggestion
{{< text yaml >}}
    serviceScopeConfigs:
      - servicesSelector:
          matchExpressions:
            - key: istio.io/global
              operator: In
              values: ["true"]
        scope: GLOBAL
{{< /text >}}
```
